sqlglot/optimizer/unnest_subqueries.py

```diff
@@ -46,20 +46,28 @@ def unnest(select, parent_select, next_alias_name):
     if not predicate or parent_select is not predicate.parent_select:
         return
 
-    # this subquery returns a scalar and can just be converted to a cross join
+    # This subquery returns a scalar and can just be converted to a cross join
     if not isinstance(predicate, (exp.In, exp.Any)):
-        having = predicate.find_ancestor(exp.Having)
         column = exp.column(select.selects[0].alias_or_name, alias)
-        if having and having.parent_select is parent_select:
+
+        clause = predicate.find_ancestor(exp.Having, exp.Where, exp.Join)
+        clause_parent_select = clause.parent_select if clause else None
+
+        if (
+            (isinstance(clause, exp.Having) and clause_parent_select is parent_select)
+            or not clause
+            or (
+                clause_parent_select is not parent_select
+                and (
+                    parent_select.args.get("group")
+                    or any(projection.find(exp.AggFunc) for projection in parent_select.selects)
+                )
+            )
+        ):
```

Comment:
This became a bit more complex, but on the plus-side we'll avoid computing `parent_select` twice for `clause`, or repeating `column = exp.Max(this=column)`. Can try to break it up into an `else` branch if it's too messy.

Reply:
Ah I noticed this is incorrect actually, let me fix that really quick..